### PR TITLE
sys/byteorder: clean up implementation

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This changes the implementation to be solely build upon `endian.h` and `unaligned.h`.

This turns `byteorder.h` basically in syntactic sugar on top of the `<endian.h>` API, reducing the complexity of the implementation and, hence, the maintenance effort.

Note that yields a small ROM reduction as well *yeah!*

```
make BOARD=nrf52840dk RIOT_CI_BUILD=1 BUILD_IN_DOCKER=1 -C tests/unittests
```

Yields before this commit:

```
   text	  data	   bss	   dec	   hex	filename
 417788	  2200	 28640	448628	 6d874	/data/riotbuild/riotbase/tests/unittests/bin/nrf52840dk/tests_unittests.elf
```

And with this commit:

```
   text	  data	   bss	   dec	   hex	filename
 417756	  2200	 28640	448596	 6d854	/data/riotbuild/riotbase/tests/unittests/bin/nrf52840dk/tests_unittests.elf
```

#### Update

This uncovered some bugs in `<endian.h>` that were not detected by the unit tests, as the unit test did not cover the magic constants but only that the right implementation was chosen. Due to the preprocessor handling undifined macros as `0` in `#if` directives, the correct implementation on little endian system was chosen anyway.

In addition to fixing the bug, the unit test was extended, so that the same bug would be detected now.

### Testing procedure

The unit tests in core cover `sys/byteorder` (historically, `byteorder.h` was once in `core`):

<details><summary><code>make BOARD=nrf52840dk BUILD_IN_DOCKER=1 -C tests/unittests tests-core flash test -j</code></summary>

```
make: Entering directory '/home/maribu/Repos/software/RIOT/byteorder/tests/unittests'
Launching build container using image "docker.io/riot/riotbuild:latest".
podman run --rm --tty --userns keep-id -v '/etc/zoneinfo/Europe/Berlin:/etc/localtime:ro' -v '/home/maribu/Repos/software/RIOT/byteorder:/data/riotbuild/riotbase:delegated' -v '/home/maribu/.cargo/registry:/data/riotbuild/.cargo/registry:delegated' -v '/home/maribu/.cargo/git:/data/riotbuild/.cargo/git:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -v '/home/maribu/.cache/RIOT:/data/riotbuild/build:delegated' -e 'BUILD_DIR=/data/riotbuild/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'    -v '/home/maribu/Repos/software/boards/riot:/data/riotbuild/external/riot:delegated' -v '/home/maribu/Repos/software/miot-pcbs/RIOT/boards:/data/riotbuild/external/boards:delegated' -v '/home/maribu/Repos/software/RIOT/master/.git:/home/maribu/Repos/software/RIOT/master/.git:delegated' -e 'BOARD=nrf52840dk' -e 'DISABLE_MODULE=auto_init auto_init_%' -e 'DEFAULT_MODULE=test_utils_interactive_sync test_utils_print_stack_usage' -e 'FEATURES_REQUIRED=' -e 'FEATURES_BLACKLIST=highlevel_stdio' -e 'FEATURES_OPTIONAL=' -e 'USEMODULE=core_mbox embunit' -e 'USEPKG='  -w '/data/riotbuild/riotbase/tests/unittests/' 'docker.io/riot/riotbuild:latest' make 'BOARD=nrf52840dk'  'EXTERNAL_BOARD_DIRS=/data/riotbuild/external/riot /data/riotbuild/external/boards' -j tests-core
Building application "tests_unittests" for "nrf52840dk" with MCU "nrf52".

"make" -C /data/riotbuild/riotbase/pkg/cmsis/ 
"make" -C /data/riotbuild/riotbase/boards/common/init
"make" -C /data/riotbuild/riotbase/boards/nrf52840dk
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/core/lib
"make" -C /data/riotbuild/riotbase/cpu/nrf52
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/tests/unittests/tests-core
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/boards/common/nrf52xxxdk
"make" -C /data/riotbuild/riotbase/sys/div
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/sys/embunit
"make" -C /data/riotbuild/riotbase/cpu/nrf52/periph
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/cpu/nrf52/vectors
"make" -C /data/riotbuild/riotbase/sys/libc
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common
"make" -C /data/riotbuild/riotbase/sys/malloc_thread_safe
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
"make" -C /data/riotbuild/riotbase/sys/test_utils/print_stack_usage
"make" -C /data/riotbuild/riotbase/sys/tsrb
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common/periph
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
   text	  data	   bss	   dec	   hex	filename
  29216	   188	  4144	 33548	  830c	/data/riotbuild/riotbase/tests/unittests/bin/nrf52840dk/tests_unittests.elf
/home/maribu/Repos/software/RIOT/byteorder/dist/tools/openocd/openocd.sh flash /home/maribu/Repos/software/RIOT/byteorder/tests/unittests/bin/nrf52840dk/tests_unittests.elf
### Flashing Target ###
Open On-Chip Debugger 0.12.0+dev-snapshot (2024-01-17-08:38)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
DEPRECATED! use 'adapter serial' not 'jlink serial'
swd
Info : J-Link OB-SAM3U128-V2-NordicSemi compiled Nov  7 2022 16:21:57
Info : Hardware version: 1.00
Info : VTarget = 3.300 V
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x2ba01477
Info : [nrf52.cpu] Cortex-M4 r0p1 processor detected
Info : [nrf52.cpu] target has 6 breakpoints, 4 watchpoints
Info : [nrf52.cpu] Examination succeed
Info : starting gdb server for nrf52.cpu on 0
Info : Listening on port 38937 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* nrf52.cpu          cortex_m   little nrf52.cpu          unknown
[nrf52.cpu] halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x000009ec msp: 0x20000200
Info : nRF52840-xxAA(build code: AA) 1024kB Flash, 256kB RAM
Warn : Adding extra erase range, 0x000072dc .. 0x00007fff
auto erase enabled
wrote 29404 bytes from file /home/maribu/Repos/software/RIOT/byteorder/tests/unittests/bin/nrf52840dk/tests_unittests.elf in 1.205779s (23.814 KiB/s)
verified 29404 bytes in 0.099144s (289.628 KiB/s)
shutdown command invoked
Done flashing
r
Due to limitations in the gcoap API it is currently not possible to use a dual stack setup
/home/maribu/Repos/software/RIOT/byteorder/dist/tools/pyterm/pyterm -p "/dev/ttyACM1" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Connect to serial port /dev/ttyACM1
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
..........................................................................................................
OK (106 tests)

make: Leaving directory '/home/maribu/Repos/software/RIOT/byteorder/tests/unittests'
```

</details>

##### tl;dr

```
READY
s
START
..........................................................................................................
OK (106 tests)
```

### Issues/PRs references

None

### CODE REVIEW

**Review by Teufelchen1 — APPROVED** [review]
🧼 🧽 Clean! ✨

File: sys/include/byteorder.h [hunk 1]
```diff
@@ -388,13 +389,6 @@ static inline uint64_t ntohll(uint64_t v);
 
 /* **************************** IMPLEMENTATION ***************************** */
 
-#ifdef HAVE_NO_BUILTIN_BSWAP16
-static inline unsigned short __builtin_bswap16(unsigned short a)
-{
-    return (a << 8) | (a >> 8);
-}
-#endif
```
[the remark this answers is not in the document]
Comment (maribu) [reply sys/include/byteorder.h:396]: For compatibility with GCC version 3 or so?

Anyway, even the AVR GCC compiler shipped in Debian / Ubuntu supports this, and that is some GCC 4.X compiler.

(We do in fact have that one in the CI. This will ensure we won't move too fast.)

[the remark this answers is not in the document]
Comment (mguetschow) [reply sys/include/byteorder.h:396]: I see, thanks!

**Comment by maribu** [comment]
This has revealed some bugs in `<endian.h>`, which have been fixed and unit tests have been added that would have caught them. Should I split the fixes out into a separate PR, or should I just update the PR description?

File: sys/libc/include/endian.h [hunk 2]
```diff
@@ -30,17 +30,17 @@ extern "C" {
...
  * @brief   A numeric constant representing big endian byte order
  */
-#define BIG_ENDIAN      implementation_defined
+#define BIG_ENDIAN      magic-number
 
 /**
  * @brief   A numeric constant representing PDP endian byte order
  */
-#define PDP_ENDIAN      implementation_defined
+#define PDP_ENDIAN       magic-number
```
Comment (Teufelchen1) [inline sys/libc/include/endian.h:43]:
```suggestion
#define PDP_ENDIAN      magic-number
```

**Review by Teufelchen1 — CHANGES_REQUESTED** [review]
I think this can stay in one PR. Just a whitespace nit left.

**Comment by maribu** [comment]
@benpicco Could you take a look at the first two commits of this PR?

I dropped the work around for the `LITTLE_ENDIAN` macro in the vendor header and just stripped that macro from the header files. We already are sanitizing the header files with a shell script for C++ compatibility, so I just extended (and renamed) that script to also handle that issue.

**Comment by maribu** [comment]
OK, this has been quite a journey, but everything is green now. I reorganized the commits into a logically sensible order. Everything should be ready for another round of review.

**Review by Teufelchen1 — APPROVED** [review]
LGTM, did not run any tests, I assume the CI is enough.

**Comment by maribu** [comment]
This time it was just a CI glitch...
